### ISSUE TITLE
Pass discount details to taxes plugins.

### ISSUE THREAD

**Issue by Akarshit** [reactioncommerce/reaction#5007]
Use case:
For an order of amount 10$, a discount of 5$ was applied. Now the consumer only has to pay the tax for 5$. 
But this information(5$ discount) is not available to the tax-plugin in the `calculateOrderTaxes` call and is also not stored on the cart either. We need to figure out a way so that `calculateOrderTaxes` can get this information.

**Comment by Akarshit** [comment]
It would be a good idea to sync up with @dancastellon before starting work on this.

**Comment by Akarshit** [comment]
@aldeed I think it would now work if the rewards data is saved in the Cart only. Because taxes are calculated even after the order has been placed(atleast the function is called again) and at that time the cart is no longer present in the database. So I can't get the rewards from the Cart.
I think it would be best if these can be added to the CommonOrder and passed to the tax-plugin.

**Comment by spencern** [comment]
@Akarshit how is this discount getting applied? Could you give some more details around the way that an order gets into this state?

**Comment by dancastellon** [comment]
@spencern SDI rewards are essentially another payment method. The issue is that until the order is actually placed, payments aren't stored on the back-end anywhere - just locally in Mobx's `cartStore`. So what I've done for rewards is add a couple mutations that SD will be implementing, that will temporarily save rewards payments to the cart under `cart.rewardsPayments`. This works when @Akarshit's tax calculation runs when there's a cart. But when the `placeOrder` mutation is called, the cart is deleted. It sounds like his tax calculation function runs again after that mutation. One solution would be to just get reward payments by filtering down the `order.payments` passed to  his `calculateOrderTaxes` like this: `order.payments.filter(payment => payment.data.method === "rewards")`.

**Comment by spencern** [comment]
I see. If the rewards are treated as a payment, why is the expectation that taxes will not apply to the order value that is paid for with rewards dollars?

**Comment by dancastellon** [comment]
@spencern I'm not sure, but I guess on the SDI side they're considered a discount, like gift cards will be.

**Comment by zenweasel** [comment]
@spencern Yes, per Bob's specs that is how they work. We have discussion this previously

**Comment by Akarshit** [comment]
For now I have started using the following logic to get the reward:
```
let cartOrder = await Cart.findOne({ _id: order.cartId }, { rewardPayments: 1 }); 
  let rewardPayments;
  if (!cartOrder) {
    // Getting rewards from Order.
    cartOrder = await Orders.findOne({ _id: order.orderId }, { payments: 1 });
    // extract the rewards and transform them into the same structure as in cart
    rewardPayments = cartOrder.payments.filter(payment => payment.method === "rewards").map((reward) => ({
      rewardNumber: reward.data.rewardNumber,
      amount: rewards.amount,
    }));
  } else {
    // Getting rewards from Cart.
    rewardPayments = cartOrder.rewardPayments;
  }
```
So this is kind of a workaround or just a "meanwhile" logic. I am reducing the priority of the issue since we have more important things to get done.

**Comment by aldeed** [comment]
@Akarshit Side comment: Should `{ rewardPayments: 1 }` be `{ projection: { rewardPayments: 1 } }`? And same for `{ payments: 1 }`.

This overall consideration should be noted somewhere for the future. When deciding whether something is a payment method or a discount, one consideration (perhaps the main) should be whether it decreases the taxable amount. Sometimes this will be at odds with the fact that it might make more sense as a payment method in the UI.

As far as the original description of this issue, there is already `totals.groupDiscountTotal` and `totals.orderDiscountTotal` on the CommonOrder passed to tax function. Is that enough to close this issue?

**Comment by spencern** [comment]
@Akarshit I noticed you've changed the impact here from `critical` to `minor` can you leave a note to explain why that was changed?

**Comment by aldeed** [comment]
I vote to close this. Contrary to this issue title and description, the discount total is already available to tax calc functions. The situation here was that something that is supposed to affect taxable amount was being treated as a payment rather than a discount in a particular RC implementation. I believe there is a working workaround where the payment details are looked up directly from the database in the tax calc function.

If we don't want to close it, then the only core change that might help would be passing the `payments` array to calculation functions, but this gets into complicated situations where payments need to know the total but also could affect the total.

**Comment by spencern** [comment]
@Akarshit I'm going with @aldeed's recommendation to close this. If you'd like to provide more information to support keeping it open, we can reopen or open a new issue.
